Add CI for PHP 8.3 and 8.4

alexander-schranz opened the pull request:
Replaces: https://github.com/Roave/infection-static-analysis-plugin/pull/497

Blocked by:

 - [x] https://github.com/vimeo/psalm/issues/11176
 - [x] https://github.com/vimeo/psalm/issues/11107

alexander-schranz committed:
Add CI for PHP 8.3 and 8.4

Ocramius commented:
@alexander-schranz sorry, I just managed to trigger CI here, and it's currently red

alexander-schranz committed:
Update dependencies

alexander-schranz commented:
@Ocramius No problem, looks like it used the lock files I did a composer update now to update the dependencies.

Not sure about the `--prefer-lowest` maybe need add some conflicts, locally I get:

> PHP Fatal error:  Cannot redeclare class PHPCSStandards\Composer\Plugin\Installers\PHPCodeSniffer\Plugin (previously declared in /Users/alexanderschranz/Documents/Projects/infection-static-analysis-plugin/vendor/dealerdirect/phpcodesniffer-composer-installer/src/Plugin.php:35) in /Users/alexanderschranz/Documents/Projects/infection-static-analysis-plugin/vendor/dealerdirect/phpcodesniffer-composer-installer/src/Plugin.php on line 35

Or do you ever stumbled over this error?

Ocramius commented:
> Or do you ever stumbled over this error?

Haven't been doing PHP for almost a year, so nope, didn't meet this error (yet) :D

alexander-schranz committed:
Fix unexpected psalm issue in internal class not matching named argument of not required to called parent construct

Ocramius commented on src/Roave/InfectionStaticAnalysis/RunStaticAnalysisAgainstEscapedMutant.php:
```diff
@@ -25,6 +25,9 @@ class RunStaticAnalysisAgainstEscapedMutant extends MutantExecutionResultFactory
 
+    /**
+     * @psalm-suppress ParamNameMismatch
+     */
```
I'd rename the parameters, fwiw

alexander-schranz replied:
The `MutantExecutionResultFactory` has a completly different constructor https://github.com/infection/infection/blob/cac7d20e5d286a37488527e477f5a695a9d7a44c/src/Mutant/MutantExecutionResultFactory.php#L53

Ocramius replied:
TIL: weird that it is flagged by SA, since constructors don't adhere to variance checks

Ocramius commented on .github/workflows/continuous-integration.yml:
```diff
@@ -73,6 +75,9 @@ jobs:
           restore-keys: "php-${{ matrix.php-version }}-${{ matrix.dependencies }}"
 
+      - name: "Remove not required dev tools"
+        run: "composer remove vimeo/psalm --no-update --no-interaction --no-progress"
```
`vimeo/psalm` is not a dev tool: it's a direct non-dev dependency

szepeviktor replied:
Is `psalm` distributed as a PHAR file?

Ocramius replied:
It's a runtime dependency, not a binary to include: it can't even be fully scoped, because we rely on symbols from its codebase.

alexander-schranz replied:
Then I think it is blocked until there is a stable release of psalm supporting 8.4:


https://github.com/vimeo/psalm/issues/11176
https://github.com/vimeo/psalm/issues/11107

alexander-schranz committed:
Remove not required dev tools

alexander-schranz committed:
Undo remove psalm

alexander-schranz committed:
Update vimeo/psalm

alexander-schranz commented:
Psalm v6 was released: https://github.com/vimeo/psalm/releases/tag/6.0.0

Ocramius commented on composer.lock:
```diff
@@ -2845,16 +2854,16 @@
         {
             "name": "vimeo/psalm",
-            "version": "5.23.1",
+            "version": "5.26.1",
```
The locked dependency still points at psalm v5, which means that we aren't running SA against v6: what happens if we only allow v6?

alexander-schranz replied:
@Ocramius do you know why the infection was fixed on a version and no version constraint was used :)?

Ocramius replied:
Yes: that's because we rely on internal API of infection that may change in any patch version.

Renovate dies regular upgrades on it.

Ocramius replied:
And indeed, the infection API did change exactly there :-P 

I'm locking down to the patch version again, as it should be.

Ocramius replied:
3e06be28f86dab8c23697245904009765827dff7

Ocramius commented:
Seems like infection itself also needs an upgrade

alexander-schranz committed:
Update dependencies

Ocramius commented on composer.json:
```diff
@@ -13,11 +13,11 @@
+        "php": "~8.1.0 || ~8.2.0 || ~8.3.0 || ~8.4.0",
         "composer-runtime-api": "^2.2",
-        "infection/infection": "0.27.11",
+        "infection/infection": "^0.27.11 || ^0.28.0 || ^0.29.0",
```
This should be an exact version, not a range.

Ocramius replied:
3e06be28f86dab8c23697245904009765827dff7

Ocramius committed:
Upgraded code to use `vimeo/psalm:^6` and `infection/infection:0.29.10`

To do this, we introduced `azjezz/psl` in the test suite, since life is too short to follow PHP-SRC's function
falsyess bullshit.

Also, various suppressions are now gone, since Psalm and Infection removed various deprecations and
mutable expressions.

The `psalm/plugin-phpunit` was also installed, so Psalm does no longer mark test classes as unused code.

Ocramius committed:
Removed PHP 8.1 from CI: no longer supported by Infection anyway

Ocramius committed:
Removed the need to instruct Psalm to re-analyze `$path`: it already does that when `$codebase` files are invalidated

Ocramius committed:
Skip running infection via `.laminas-ci.json`, since it can't use `phpdbg`

Fixes:

```
Error: ] Project tests must be in a passing state before running Infection.
         Infection runs the test suite in a RANDOM order. Make sure your tests
         do not have hidden dependencies.

         You can add these attributes to `phpunit.xml` to check it: <phpunit
         executionOrder="random" resolveDependencies="true" ...

         If you don't want to let Infection run tests in a random order, set the
         `executionOrder` to some value, for example <phpunit
         executionOrder="default"

         Check the executed command to identify the problem:
         '/usr/bin/phpdbg8.2' '-qrr' '/github/workspace/vendor/bin/phpunit'
         '--configuration'
         '/tmp/infection/phpunitConfiguration.initial.infection.xml'
         '--coverage-xml=/tmp/infection/coverage-xml'
         '--log-junit=/tmp/infection/junit.xml'
         PHPUnit reported an exit code of 1.
         Refer to the PHPUnit's output below:
         STDOUT:
         PHPUnit 10.5.41 by Sebastian Bergmann and contributors.

         Runtime:       PHP 8.2.24
         Configuration:
         /tmp/infection/phpunitConfiguration.initial.infection.xml
         Random Seed:   1737943321

         There was 1 PHPUnit test runner warning:

         1) No code coverage driver available

         No tests executed!
```

Ocramius commented:
Shipping this after few adjustments then: thanks for the help, @alexander-schranz!

alexander-schranz commented:
@Ocramius thank you!
